apps/platform/src/app/(main)/(project)/[workspace]/[project]/@environment/page.tsx

```diff
@@ -18,33 +18,47 @@ import ControllerInstance from '@/lib/controller-instance'
 import { Button } from '@/components/ui/button'
 import { useHttp } from '@/hooks/use-http'
 
-const INITIAL_DISPLAY_COUNT = 10
-
 function EnvironmentPage(): React.JSX.Element {
   const setIsCreateEnvironmentOpen = useSetAtom(createEnvironmentOpenAtom)
   const isDeleteEnvironmentOpen = useAtomValue(deleteEnvironmentOpenAtom)
   const isEditEnvironmentOpen = useAtomValue(editEnvironmentOpenAtom)
   const [environments, setEnvironments] = useAtom(environmentsOfProjectAtom)
   const selectedProject = useAtomValue(selectedProjectAtom)
   const selectedEnvironment = useAtomValue(selectedEnvironmentAtom)
-  const [visibleCount, setVisibleCount] = useState<number>(INITIAL_DISPLAY_COUNT)
+  const [page, setPage] = useState(0)
+  const [hasMore, setHasMore] = useState(true)
+  const [loading, setLoading] = useState(false)
 
   const getAllEnvironmentsOfProject = useHttp(() =>
     ControllerInstance.getInstance().environmentController.getAllEnvironmentsOfProject(
       {
-        projectSlug: selectedProject!.slug
+        projectSlug: selectedProject!.slug,
+        page,
       }
     )
   )
 
   useEffect(() => {
-    selectedProject &&
-      getAllEnvironmentsOfProject().then(({ data, success }) => {
+    if (!selectedProject) return 
+    setLoading(true)
+    getAllEnvironmentsOfProject()
+      .then(({ data, success }) => {
         if (success && data) {
-          setEnvironments(data.items)
+          const newData = page === 0 ? data.items : [...environments, ...data.items];
+          setEnvironments(newData);
+          if (newData.length >= data.metadata.totalCount) setHasMore(false);
         }
       })
-  }, [getAllEnvironmentsOfProject, selectedProject, setEnvironments])
+      .finally(() => {
+        setLoading(false)
+      })
+  }, [selectedProject, page, getAllEnvironmentsOfProject, setEnvironments])
+
+  const handlePageShift = () => {
```

Comment:
I feel you should extract all of the logic in the use effect hook to a separate function.
- call that function once in use effect
- call that function whenever load more is clicked